### ISSUE TITLE
Improve built in scripts

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#4890]
### Describe the project you are working on

Spaceship game

### Describe the problem or limitation you are having in your project

- Built in scripts can be removed breaking implementation down the road
- Built in scripts can extend a different node when extended breaking implementation 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Built in script should be treated like part of the node or scene.
This means built in scripts would implicitly automatically implement extends <inherited builtin script type>

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

If a scene is extended from a scene with a built in script, the script in the extended scene would..
- Have a greyed out icon that could be clicked to extend the script
- Wouldn't have add or remove options
- Wouldn't need `extends` because it can be deferred from the node and can break if it is changed to something else.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It would be used to implement logic inside a node without creating a separate file.

### Is there a reason why this should be core and not an add-on in the asset library?

Built in scripts are already a part of core, this just makes them act similar to game maker objects where you can extend the object without breaking implementation.

It also is much easier to use and manage the normal scripts if it doesn't have bugs

**Comment by Shadowblitz16** [comment]
I had a suggestion about forcing the node to extend the type of node it's attached to. maybe that would be a better implementation 
of type security.

**Comment by Shadowblitz16** [comment]
If anybody wondering what I mean by breaking implementation...

The idea behind this is there would be a possibility of a hard 1:1 relationship with a node and it's script.

If you have a node that calls on a member of another node you want to make sure its there.
This combined with static typing allows this.

Right now you can't ensure a node has a member because the script can be removed and in the 99% of cases of built in scripts your not going to remove it anyways.

**Comment by dalexeev** [comment]
Is the problem that you personally quite often accidentally delete the script and do not notice it, or is the problem that there is a fundamental possibility to do this? If the latter, then it seems to me that this is the case when we can say "just don't do bad things", since it is impossible to provide unbreakable guarantees.

But I agree that in some places it would be nice to improve the UI of the editor (for example, in the case of inherited scenes, instead of the Attach/Remove button, it should be Extend/Unextend).

**Comment by Shadowblitz16** [comment]
I think it would make the editor nicer definitely.

I don't think we should prevent it overall it might be useful to override them in a plugin, but as far as the editor goes it makes the editor nicer and things easier to understand and use.

I hope one day Godot will have some sort of live edit where people can design games together over the network.
In that case you wouldn't want someone attaching a script to a node that already has a one defined in a previous scene.

I still stick to my beliefs that no script should be completely replaced if it has been inherited from another scene but in the case of built in scripts it really doesn't make sense.

**Comment by zinnschlag** [comment]
Nope. Simple counter example: You extend a scene (which has a script). Then you want to add an existing script to the scene (which happens to extend the correct script). Without an add button, not possible.

Also not okay with removing the remove (clear) function. Yes, removing may temporarily break the scene. That's okay. When doing refactoring I would rather have a temporarily broken scene (which will happen anyway) than having to dance around such a limitation.

**Comment by dalexeev** [comment]
I don't see the need (and don't think it's right) to make a distinction here between built-in and normal scripts.

> But I agree that in some places it would be nice to improve the UI of the editor (for example, in the case of inherited scenes, instead of the Attach/Remove button, it should be Extend/Unextend).

Mock-up:

![](https://user-images.githubusercontent.com/47700418/179388297-c4b4dc88-af15-4379-a5d7-a18900e58fbf.png)

**Comment by zinnschlag** [comment]
@dalexeev Funny. It seems you argued against such a feature in #2670. Changed your mind? I suggest you read your arguments in #2670 again because they were very good, At least for the case of non built-in scripts. Can't competently comment on built-in scripts since I don't use them and therefore have no experience with them.

Also, your new proposal misses the ability to replace the script of an inherited scene with a pre-existing script (after all, it is possible to use the same script for multiple scenes).

**Comment by dalexeev** [comment]
@zinnschlag 

> Is the problem that you personally quite often accidentally delete the script and do not notice it, or is the problem that there is a fundamental possibility to do this?

I think changes to the UI are a good idea. But we still cannot provide strict guarantees.

**Comment by zinnschlag** [comment]
IMHO little value and additional complexity in the UI. Guess we have to agree to disagree here.

> Clicking the Extend Script button should open a create/select script dialog where you can select an existing script. The editor checks that the new/existing script inherits the base scene script.

That sounds highly unintuitive. Seeing "extend script" I would not think of the ability to load a pre-existing script. But it seems that is what Godot is doing in its current version (only just discovered this feature after you mentioned it). Not a good design IMO.

**Comment by dalexeev** [comment]
> That sounds highly unintuitive. Seeing "extend script" I would not think of the ability to load a pre-existing script.

The Extend Script button **already** works this way.

**Comment by zinnschlag** [comment]
That's what I said ;)

**Comment by dalexeev** [comment]
Just now I realized that Attach Script (Replace Script on my mock-up) and Extend Script are the same dialog (only with different initial values, plus the Inherits field is disabled) if the Extend Script button was pressed. That is, in the case of legacy scenes, it makes sense to display the Replace Script button as well - you will be able to select any script, however, if it does not inherit the script of the base scene, a warning will be displayed.

In #2670, I wrote that this change would require too much effort. However, if we limit ourselves to editor changes, then it seems to me that not too much code is required. Admittedly, the UI really needs some tweaking, as right now you can't quickly reset the script to the base scene script, you have to detach it first and then re-attach the base scene script by finding it in the file system. From this point of view, I really changed my mind.

I also had an idea to move this to the Change Type context menu item (display all scripts in the class selection dialog) and create the illusion of full-fledged classes, but then I thought that this was not a good idea.

**Comment by Shadowblitz16** [comment]
> I don't see the need (and don't think it's right) to make a distinction here between built-in and normal scripts.
> 
> > But I agree that in some places it would be nice to improve the UI of the editor (for example, in the case of inherited scenes, instead of the Attach/Remove button, it should be Extend/Unextend).
> 
> Mock-up:
> 
> ![](https://user-images.githubusercontent.com/47700418/179388676-25c8570d-04d8-44fc-9e56-7194e5ef0037.png)

I like this very much but one thing I would change is the script icon on the node (not the one in the header) would be greyed out if it was from a derived scene, unless extended in that scene. 

The same icon that used to be used for plugins in 3.X.X

**Comment by Shadowblitz16** [comment]
@dalexeev You may want to make a proposal for your idea.
